README.md
<h1 align="center">
	<img
		width="300"
		alt="aRp Systems"
		src="https://raw.githubusercontent.com/arpsystems/arpsystems/master/client/img/logo-vertical-transparent-bg.svg?sanitize=true">
</h1>

<h3 align="center">
	Modern web IRC client designed for self-hosting
</h3>

<p align="center">
	<strong>
		<a href="https://arp.systems/">Website</a>
		•
		<a href="https://arp.systems/docs">Docs</a>
		•
		<a href="https://demo.arp.systems/">Demo</a>
	</strong>
</p>
<p align="center">
	<a href="https://demo.arp.systems/"><img
		alt="#arpsystems IRC channel on freenode"
		src="https://img.shields.io/badge/freenode-%23arpsystems-415364.svg?colorA=ff9e18&style=flat-square"></a>
	<br>
	<a href="https://yarn.pm/arpsystems"><img
		alt="npm version"
		src="https://img.shields.io/npm/v/arpsystems.svg?style=flat-square&maxAge=3600"></a>
	<a href="https://travis-ci.com/arpsystems/arpsystems"><img
		alt="Travis CI Build Status"
		src="https://img.shields.io/travis/com/arpsystems/arpsystems/master.svg?style=flat-square&maxAge=60"></a>
	<a href="https://david-dm.org/arpsystems/arpsystems"><img
		alt="Dependencies Status"
		src="https://img.shields.io/david/arpsystems/arpsystems.svg?style=flat-square&maxAge=3600"></a>
	<a href="https://npm-stat.com/charts.html?package=arpsystems&from=2016-02-12"><img
		alt="Total downloads on npm"
		src="https://img.shields.io/npm/dt/arpsystems.svg?colorB=007dc7&style=flat-square&maxAge=3600"></a>
</p>

<p align="center">
	<img src="https://raw.githubusercontent.com/arpsystems/arpsystems.github.io/master/img/arpsystems-screenshot.png" width="550">
</p>

## Overview

* **Modern features brought to IRC.** Push notifications, link previews, new message markers, and more bring IRC to the 21st century.
* **Always connected.** Remains connected to IRC servers while you are offline.
* **Cross platform.** It doesn't matter what OS you use, it just works wherever Node.js runs.
* **Responsive interface.** The client works smoothly on every desktop, smartphone and tablet.
* **Synchronized experience.** Always resume where you left off no matter what device.

To learn more about configuration, usage and features of The Lounge, take a look at [the website](https://arp.systems).

The Lounge is the official and community-managed fork of [Shout](https://github.com/erming/shout), by [Mattias Erming](https://github.com/erming).

## Installation and usage

The Lounge requires [Node.js](https://nodejs.org/) v6 or more recent.
[Yarn package manager](https://yarnpkg.com/) is also recommended.  
If you want to install with npm, `--unsafe-perm` is required for a correct install.

### Running stable releases using Yarn (recommended)

Run this in a terminal to install (or upgrade) the latest stable release from
[npm registry](https://www.npmjs.com/):

```sh
yarn global add arpsystems
```

If you already have The Lounge installed globally, use the following command to update it:

```sh
yarn global upgrade arpsystems
```

When installation is complete, run:

```sh
arpsystems start
```

For more information, read the [usage documentation](https://arp.systems/docs/usage) or run:

```sh
arpsystems --help
```

### Running from source

The following commands install and run the development version of The Lounge:

```sh
git clone https://github.com/arpsystems/arpsystems.git
cd arpsystems
yarn install
NODE_ENV=production yarn build
yarn start
```

When installed like this, `arpsystems` executable is not created. Use `node index <command>` to run commands.

⚠️ While it is the most recent codebase, this is not production-ready! Run at
your own risk. It is also not recommended to run this as root.

## Development setup

Simply follow the instructions to run The Lounge from source above, on your own
fork.

Before submitting any change, make sure to:

- Read the [Contributing instructions](https://github.com/arpsystems/arpsystems/blob/master/.github/CONTRIBUTING.md#contributing)
- Run `yarn test` to execute linters and test suite
- Run `yarn build` if you change or add anything in `client/js` or `client/views`
- `yarn dev` can be used to start The Lounge and watch for any file changes in the client folder


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

<!-- New entries go after this line -->

## v3.0.1 - 2019-02-11

This patch release fixes a few small things that made it to the previous release.

It includes fixes for issues related to: reconnecting when a nickname is already in use, unnecessary highlights on invites, the auto-expandable message input, channels starting with `&` or `+`, the ` /disconnect` command, and file uploads. Additionnally, `part` messages for the current user are now logged.

A more comprehensive list of changes is available below.

### Changed

- Update production dependencies to their latest versions:
  - `file-type` ([#2939](https://github.com/arpsystems/arpsystems/pull/2939), [#3000](https://github.com/arpsystems/arpsystems/pull/3000))
  - `chalk` ([#2957](https://github.com/arpsystems/arpsystems/pull/2957))
  - `yarn` ([#2965](https://github.com/arpsystems/arpsystems/pull/2965))
  - `socketio-file-upload` ([#3001](https://github.com/arpsystems/arpsystems/pull/3001))
  - `irc-framework` ([#3040](https://github.com/arpsystems/arpsystems/pull/3040))

### Fixed

- Always emit `part` message (required for logging) ([#2989](https://github.com/arpsystems/arpsystems/pull/2989) by [@xPaw](https://github.com/xPaw))
- Fix `line-height` to match height in input ([#2995](https://github.com/arpsystems/arpsystems/pull/2995) by [@xPaw](https://github.com/xPaw))
- Fix incorrect 404 error for files with unknown file type ([#3010](https://github.com/arpsystems/arpsystems/pull/3010) by [@xPaw](https://github.com/xPaw))
- Fix up link insertion after uploads to be saner ([#3011](https://github.com/arpsystems/arpsystems/pull/3011) by [@xPaw](https://github.com/xPaw))
- Do not get highlight on invites ([#3031](https://github.com/arpsystems/arpsystems/pull/3031) by [@creesch](https://github.com/creesch))
- Allow forcing /disconnect to stop reconnection timer from running ([#3034](https://github.com/arpsystems/arpsystems/pull/3034) by [@xPaw](https://github.com/xPaw))
- Fix up textarea growing to avoid rounding issues in Chrome ([#3033](https://github.com/arpsystems/arpsystems/pull/3033) by [@xPaw](https://github.com/xPaw))

### Documentation

In the main repository:

- Do not report the (renamed) Renovate bot as a contributor ([#2997](https://github.com/arpsystems/arpsystems/pull/2997) by [@astorije](https://github.com/astorije))
- Add `webirc` key and commas in configuration file examples to make them more understandable ([#3008](https://github.com/arpsystems/arpsystems/pull/3008) by [@xPaw](https://github.com/xPaw))

On the [website repository](https://github.com/arpsystems/arpsystems.github.io):

- Use new URL for `ansible-arpsystems` ([#196](https://github.com/arpsystems/arpsystems.github.io/pull/196) by [@astorije](https://github.com/astorije))
- Add a link to a blog post in the Community page ([#195](https://github.com/arpsystems/arpsystems.github.io/pull/195) by [@astorije](https://github.com/astorije))
- Simplify Caddy reverse proxy example by using `transparent` shorthand ([#197](https://github.com/arpsystems/arpsystems.github.io/pull/197) by [@MarkOtzen](https://github.com/MarkOtzen))
- Place config update comment in noteblock for visual clarity. ([#198](https://github.com/arpsystems/arpsystems.github.io/pull/198) by [@EntityReborn](https://github.com/EntityReborn))
- Regenerate config ([`78ada86`](https://github.com/arpsystems/arpsystems.github.io/commit/78ada860206376116fcdda13e9c09399e811bfde) by [@xPaw](https://github.com/xPaw))
- Add title to iframe ([`7c4d0d2`](https://github.com/arpsystems/arpsystems.github.io/commit/7c4d0d279a686a6db9c0352a0f7c8f002a7678b0) by [@xPaw](https://github.com/xPaw))
- Apply `word-break` to `<code>` only on small screens ([#202](https://github.com/arpsystems/arpsystems.github.io/pull/202) by [@xPaw](https://github.com/xPaw))

### Internals

- Use `require.resolve` for yarn ([#2993](https://github.com/arpsystems/arpsystems/pull/2993) by [@xPaw](https://github.com/xPaw))
- Keep using npm to clean up dist-tags after a stable release ([#2999](https://github.com/arpsystems/arpsystems/pull/2999) by [@astorije](https://github.com/astorije))
- Bump Travis to use Node 10 LTS for Windows, OSX, and production builds ([#3003](https://github.com/arpsystems/arpsystems/pull/3003) by [@xPaw](https://github.com/xPaw))
- Make sure the editorconfig indent size applies to all file formats ([#3022](https://github.com/arpsystems/arpsystems/pull/3022) by [@astorije](https://github.com/astorije))
- Update development dependencies to their latest versions:
  - `@babel/preset-env` ([#2942](https://github.com/arpsystems/arpsystems/pull/2942))
  - `moment` ([#2991](https://github.com/arpsystems/arpsystems/pull/2991))
  - `primer-tooltips` ([#2962](https://github.com/arpsystems/arpsystems/pull/2962), [#3001](https://github.com/arpsystems/arpsystems/pull/3001))
  - `emoji-regex` ([#2941](https://github.com/arpsystems/arpsystems/pull/2941))
  - `handlebars-loader` ([#2938](https://github.com/arpsystems/arpsystems/pull/2938), [#3001](https://github.com/arpsystems/arpsystems/pull/3001))
  - `@fortawesome/fontawesome-free` ([#2944](https://github.com/arpsystems/arpsystems/pull/2944), [#3006](https://github.com/arpsystems/arpsystems/pull/3006), [#3016](https://github.com/arpsystems/arpsystems/pull/3016))
  - `sinon` ([#2986](https://github.com/arpsystems/arpsystems/pull/2986))
  - `babel-loader` ([#2954](https://github.com/arpsystems/arpsystems/pull/2954))

webpack.config.js
"use strict";

const webpack = require("webpack");
const path = require("path");
const CopyPlugin = require("copy-webpack-plugin");
const MiniCssExtractPlugin = require("mini-css-extract-plugin");
const VueLoaderPlugin = require("vue-loader/lib/plugin");

const config = {
	mode: process.env.NODE_ENV === "production" ? "production" : "development",
	entry: {
		"js/bundle.js": path.resolve(__dirname, "client/js/arp.js"),
		"css/style": path.resolve(__dirname, "client/css/style.css"),
	},
	devtool: "source-map",
	output: {
		path: path.resolve(__dirname, "public"),
		filename: "[name]",
		publicPath: "/",
	},
	module: {
		rules: [
			{
				test: /\.vue$/,
				use: {
					loader: "vue-loader",
					options: {
						compilerOptions: {
							preserveWhitespace: false,
						},
					},
				},
			},
			{
				test: /\.css$/,
				include: [
					path.resolve(__dirname, "client"),
				],
				use: [
					MiniCssExtractPlugin.loader,
					{
						loader: "css-loader",
						options: {
							url: false,
							minimize: process.env.NODE_ENV === "production",
						},
					},
				],
			},
			{
				test: /\.js$/,
				include: [
					path.resolve(__dirname, "client"),
				],
				use: {
					loader: "babel-loader",
					options: {
						presets: [
							["@babel/env", {
								targets: {
									browsers: [
										"last 1 year",
										"firefox esr",
									],
								},
							}],
						],
					},
				},
			},
			{
				test: /\.tpl$/,
				include: [
					path.resolve(__dirname, "client/views"),
				],
				use: [{
					loader: "handlebars-loader",
					options: {
						helperDirs: [
							path.resolve(__dirname, "client/js/libs/handlebars"),
						],
						extensions: [
							".tpl",
						],
					},
				}, {
					loader: "html-minifier-loader",
					options: {
						ignoreCustomFragments: [
							/{{[\s\S]*?}}/,
						],
					},
				}],
			},
		],
	},
	optimization: {
		splitChunks: {
			cacheGroups: {
				commons: {
					test: /[\\/]node_modules[\\/]/,
					name: "js/bundle.vendor.js",
					chunks: "all",
				},
			},
		},
	},
	externals: {
		json3: "JSON", // socket.io uses json3.js, but we do not target any browsers that need it
	},
	plugins: [
		new MiniCssExtractPlugin(),
		new VueLoaderPlugin(),
		new CopyPlugin([
			{
				from: "./node_modules/@fortawesome/fontawesome-free/webfonts/fa-solid-900.woff*",
				to: "fonts/[name].[ext]",
			},
			{
				from: "./client/js/loading-error-handlers.js",
				to: "js/[name].[ext]",
			},
			{
				from: "./client/*",
				to: "[name].[ext]",
				ignore: "index.html.tpl",
			},
			{
				from: "./client/audio/*",
				to: "audio/[name].[ext]",
			},
			{
				from: "./client/img/*",
				to: "img/[name].[ext]",
			},
			{
				from: "./client/themes/*",
				to: "themes/[name].[ext]",
			},
			{
				from: "./node_modules/primer-tooltips/build/build.css",
				to: "css/primer-tooltips.[ext]",
			},
		]),
		new webpack.IgnorePlugin(/^\.\/locale$/, /moment$/),
		// socket.io uses debug, we don't need it
		new webpack.NormalModuleReplacementPlugin(/debug/, path.resolve(__dirname, "scripts/noop.js")),
	],
};

module.exports = config;


src/plugins/webpush.js
"use strict";

const _ = require("lodash");
const log = require("../log");
const fs = require("fs");
const path = require("path");
const WebPushAPI = require("web-push");
const Helper = require("../helper");

class WebPush {
	constructor() {
		const vapidPath = path.join(Helper.getHomePath(), "vapid.json");

		if (fs.existsSync(vapidPath)) {
			const data = fs.readFileSync(vapidPath, "utf-8");
			const parsedData = JSON.parse(data);

			if (typeof parsedData.publicKey === "string" && typeof parsedData.privateKey === "string") {
				this.vapidKeys = {
					publicKey: parsedData.publicKey,
					privateKey: parsedData.privateKey,
				};
			}
		}

		if (!this.vapidKeys) {
			this.vapidKeys = WebPushAPI.generateVAPIDKeys();

			fs.writeFileSync(vapidPath, JSON.stringify(this.vapidKeys, null, "\t"));

			log.info("New VAPID key pair has been generated for use with push subscription.");
		}

		WebPushAPI.setVapidDetails(
			"https://github.com/arpsystems/arpsystems",
			this.vapidKeys.publicKey,
			this.vapidKeys.privateKey
		);
	}

	push(client, payload, onlyToOffline) {
		_.forOwn(client.config.sessions, ({pushSubscription}, token) => {
			if (pushSubscription) {
				if (onlyToOffline && _.find(client.attachedClients, {token}) !== undefined) {
					return;
				}

				this.pushSingle(client, pushSubscription, payload);
			}
		});
	}

	pushSingle(client, subscription, payload) {
		WebPushAPI
			.sendNotification(subscription, JSON.stringify(payload))
			.catch((error) => {
				if (error.statusCode >= 400 && error.statusCode < 500) {
					log.warn(`WebPush subscription for ${client.name} returned an error (${error.statusCode}), removing subscription`);

					_.forOwn(client.config.sessions, ({pushSubscription}, token) => {
						if (pushSubscription && pushSubscription.endpoint === subscription.endpoint) {
							client.unregisterPushSubscription(token);
						}
					});

					return;
				}

				log.error(`WebPush Error (${error})`);
			});
	}
}

module.exports = WebPush;


src/server.js
"use strict";

const _ = require("lodash");
const log = require("./log");
const pkg = require("../package.json");
const Client = require("./client");
const ClientManager = require("./clientManager");
const express = require("express");
const fs = require("fs");
const path = require("path");
const io = require("socket.io");
const dns = require("dns");
const Uploader = require("./plugins/uploader");
const Helper = require("./helper");
const colors = require("chalk");
const net = require("net");
const Identification = require("./identification");
const changelog = require("./plugins/changelog");

const themes = require("./plugins/packages/themes");
themes.loadLocalThemes();

const packages = require("./plugins/packages/index");
packages.loadPackages();

// The order defined the priority: the first available plugin is used
// ALways keep local auth in the end, which should always be enabled.
const authPlugins = [
	require("./plugins/auth/ldap"),
	require("./plugins/auth/local"),
];

// A random number that will force clients to reload the page if it differs
const serverHash = Math.floor(Date.now() * Math.random());

let manager = null;

module.exports = function() {
	log.info(`aRp systems ${colors.green(Helper.getVersion())} \
(Node.js ${colors.green(process.versions.node)} on ${colors.green(process.platform)} ${process.arch})`);
	log.info(`Configuration file: ${colors.green(Helper.getConfigPath())}`);

	const staticOptions = {
		redirect: false,
		maxAge: 86400 * 1000,
	};

	const app = express()
		.set("env", "production")
		.disable("x-powered-by")
		.use(allRequests)
		.use(index)
		.use(express.static(path.join(__dirname, "..", "public"), staticOptions))
		.use("/storage/", express.static(Helper.getStoragePath(), staticOptions));

	if (Helper.config.fileUpload.enable) {
		Uploader.router(app);
	}

	// This route serves *installed themes only*. Local themes are served directly
	// from the `public/themes/` folder as static assets, without entering this
	// handler. Remember this if you make changes to this function, serving of
	// local themes will not get those changes.
	app.get("/themes/:theme.css", (req, res) => {
		const themeName = req.params.theme;
		const theme = themes.getFilename(themeName);

		if (theme === undefined) {
			return res.status(404).send("Not found");
		}

		return res.sendFile(theme);
	});

	app.get("/packages/:package/:filename", (req, res) => {
		const packageName = req.params.package;
		const fileName = req.params.filename;
		const packageFile = packages.getPackage(packageName);

		if (!packageFile || !packages.getStylesheets().includes(`${packageName}/${fileName}`)) {
			return res.status(404).send("Not found");
		}

		const packagePath = Helper.getPackageModulePath(packageName);
		return res.sendFile(path.join(packagePath, fileName));
	});

	let server = null;

	if (Helper.config.public && (Helper.config.ldap || {}).enable) {
		log.warn("Server is public and set to use LDAP. Set to private mode if trying to use LDAP authentication.");
	}

	if (!Helper.config.https.enable) {
		server = require("http");
		server = server.createServer(app);
	} else {
		const keyPath = Helper.expandHome(Helper.config.https.key);
		const certPath = Helper.expandHome(Helper.config.https.certificate);
		const caPath = Helper.expandHome(Helper.config.https.ca);

		if (!keyPath.length || !fs.existsSync(keyPath)) {
			log.error("Path to SSL key is invalid. Stopping server...");
			process.exit();
		}

		if (!certPath.length || !fs.existsSync(certPath)) {
			log.error("Path to SSL certificate is invalid. Stopping server...");
			process.exit();
		}

		if (caPath.length && !fs.existsSync(caPath)) {
			log.error("Path to SSL ca bundle is invalid. Stopping server...");
			process.exit();
		}

		server = require("https");
		server = server.createServer({
			key: fs.readFileSync(keyPath),
			cert: fs.readFileSync(certPath),
			ca: caPath ? fs.readFileSync(caPath) : undefined,
		}, app);
	}

	let listenParams;

	if (typeof Helper.config.host === "string" && Helper.config.host.startsWith("unix:")) {
		listenParams = Helper.config.host.replace(/^unix:/, "");
	} else {
		listenParams = {
			port: Helper.config.port,
			host: Helper.config.host,
		};
	}

	server.on("error", (err) => log.error(`${err}`));

	server.listen(listenParams, () => {
		if (typeof listenParams === "string") {
			log.info("Available on socket " + colors.green(listenParams));
		} else {
			const protocol = Helper.config.https.enable ? "https" : "http";
			const address = server.address();

			log.info(
				"Available at " +
				colors.green(`${protocol}://${address.address}:${address.port}/`) +
				` in ${colors.bold(Helper.config.public ? "public" : "private")} mode`
			);
		}

		const sockets = io(server, {
			wsEngine: "ws",
			serveClient: false,
			transports: Helper.config.transports,
		});

		sockets.on("connect", (socket) => {
			if (Helper.config.public) {
				performAuthentication.call(socket, {});
			} else {
				socket.emit("auth", {
					serverHash: serverHash,
					success: true,
				});
				socket.on("auth", performAuthentication);
			}
		});

		manager = new ClientManager();

		new Identification((identHandler) => {
			manager.init(identHandler, sockets);
		});

		// Handle ctrl+c and kill gracefully
		let suicideTimeout = null;

		const exitGracefully = function() {
			if (suicideTimeout !== null) {
				return;
			}

			log.info("Exiting...");

			// Close all client and IRC connections
			manager.clients.forEach((client) => client.quit());

			if (Helper.config.prefetchStorage) {
				log.info("Clearing prefetch storage folder, this might take a while...");

				require("./plugins/storage").emptyDir();
			}

			// Forcefully exit after 3 seconds
			suicideTimeout = setTimeout(() => process.exit(1), 3000);

			// Close http server
			server.close(() => {
				clearTimeout(suicideTimeout);
				process.exit(0);
			});
		};

		process.on("SIGINT", exitGracefully);
		process.on("SIGTERM", exitGracefully);

		// Clear storage folder after server starts successfully
		if (Helper.config.prefetchStorage) {
			require("./plugins/storage").emptyDir();
		}
	});

	return server;
};

function getClientLanguage(socket) {
	const acceptLanguage = socket.handshake.headers["accept-language"];

	if (typeof acceptLanguage === "string" && /^[\x00-\x7F]{1,50}$/.test(acceptLanguage)) {
		// only allow ASCII strings between 1-50 characters in length
		return acceptLanguage;
	}

	return null;
}

function getClientIp(socket) {
	let ip = socket.handshake.address || "127.0.0.1";

	if (Helper.config.reverseProxy) {
		const forwarded = (socket.request.headers["x-forwarded-for"] || "").split(/\s*,\s*/).filter(Boolean);

		if (forwarded.length && net.isIP(forwarded[0])) {
			ip = forwarded[0];
		}
	}

	return ip.replace(/^::ffff:/, "");
}

function allRequests(req, res, next) {
	res.setHeader("X-Content-Type-Options", "nosniff");
	return next();
}

function index(req, res, next) {
	if (req.url.split("?")[0] !== "/") {
		return next();
	}

	const policies = [
		"default-src 'none'", // default to nothing
		"form-action 'self'", // 'self' to fix saving passwords in Firefox, even though login is handled in javascript
		"connect-src 'self' ws: wss:", // allow self for polling; websockets
		"style-src 'self' https: 'unsafe-inline'", // allow inline due to use in irc hex colors
		"script-src 'self'", // javascript
		"worker-src 'self'", // service worker
		"child-src 'self'", // deprecated fall back for workers, Firefox <58, see #1902
		"manifest-src 'self'", // manifest.json
		"font-src 'self' https:", // allow loading fonts from secure sites (e.g. google fonts)
		"media-src 'self' https:", // self for notification sound; allow https media (audio previews)
	];

	// If prefetch is enabled, but storage is not, we have to allow mixed content
	// - https://user-images.githubusercontent.com is where we currently push our changelog screenshots
	// - data: is required for the HTML5 video player
	if (Helper.config.prefetchStorage || !Helper.config.prefetch) {
		policies.push("img-src 'self' data: https://user-images.githubusercontent.com");
		policies.unshift("block-all-mixed-content");
	} else {
		policies.push("img-src http: https: data:");
	}

	res.setHeader("Content-Type", "text/html");
	res.setHeader("Content-Security-Policy", policies.join("; "));
	res.setHeader("Referrer-Policy", "no-referrer");

	return fs.readFile(path.join(__dirname, "..", "client", "index.html.tpl"), "utf-8", (err, file) => {
		if (err) {
			throw err;
		}

		res.send(_.template(file)(getServerConfiguration()));
	});
}

function initializeClient(socket, client, token, lastMessage) {
	socket.emit("authorized");

	client.clientAttach(socket.id, token);

	if (Helper.config.fileUpload.enable) {
		new Uploader(socket);
	}

	socket.on("disconnect", function() {
		process.nextTick(() => client.clientDetach(socket.id));
	});

	socket.on("input", (data) => {
		if (typeof data === "object") {
			client.input(data);
		}
	});

	socket.on("more", (data) => {
		if (typeof data === "object") {
			const history = client.more(data);

			if (history !== null) {
				socket.emit("more", history);
			}
		}
	});

	socket.on("network:new", (data) => {
		if (typeof data === "object") {
			// prevent people from overriding webirc settings
			data.ip = null;
			data.hostname = null;
			data.uuid = null;
			data.commands = null;
			data.ignoreList = null;

			client.connect(data);
		}
	});

	socket.on("network:get", (data) => {
		if (typeof data !== "string") {
			return;
		}

		const network = _.find(client.networks, {uuid: data});

		if (!network) {
			return;
		}

		socket.emit("network:info", getClientConfiguration(network.export()));
	});

	socket.on("network:edit", (data) => {
		if (typeof data !== "object") {
			return;
		}

		const network = _.find(client.networks, {uuid: data.uuid});

		if (!network) {
			return;
		}

		network.edit(client, data);
	});

	if (!Helper.config.public && !Helper.config.ldap.enable) {
		socket.on("change-password", (data) => {
			if (typeof data === "object") {
				const old = data.old_password;
				const p1 = data.new_password;
				const p2 = data.verify_password;

				if (typeof p1 === "undefined" || p1 === "") {
					socket.emit("change-password", {
						error: "Please enter a new password",
					});
					return;
				}

				if (p1 !== p2) {
					socket.emit("change-password", {
						error: "Both new password fields must match",
					});
					return;
				}

				Helper.password
					.compare(old || "", client.config.password)
					.then((matching) => {
						if (!matching) {
							socket.emit("change-password", {
								error: "The current password field does not match your account password",
							});
							return;
						}

						const hash = Helper.password.hash(p1);

						client.setPassword(hash, (success) => {
							const obj = {};

							if (success) {
								obj.success = "Successfully updated your password";
							} else {
								obj.error = "Failed to update your password";
							}

							socket.emit("change-password", obj);
						});
					}).catch((error) => {
						log.error(`Error while checking users password. Error: ${error}`);
					});
			}
		});
	}

	socket.on("open", (data) => {
		client.open(socket.id, data);
	});

	socket.on("sort", (data) => {
		if (typeof data === "object") {
			client.sort(data);
		}
	});

	socket.on("names", (data) => {
		if (typeof data === "object") {
			client.names(data);
		}
	});

	socket.on("changelog", () => {
		changelog.fetch((data) => {
			socket.emit("changelog", data);
		});
	});

	socket.on("msg:preview:toggle", (data) => {
		if (typeof data !== "object") {
			return;
		}

		const networkAndChan = client.find(data.target);

		if (!networkAndChan) {
			return;
		}

		const message = networkAndChan.chan.findMessage(data.msgId);

		if (!message) {
			return;
		}

		const preview = message.findPreview(data.link);

		if (preview) {
			preview.shown = data.shown;
		}
	});

	if (!Helper.config.public) {
		socket.on("push:register", (subscription) => {
			if (!client.config.sessions.hasOwnProperty(token)) {
				return;
			}

			const registration = client.registerPushSubscription(client.config.sessions[token], subscription);

			if (registration) {
				client.manager.webPush.pushSingle(client, registration, {
					type: "notification",
					timestamp: Date.now(),
					title: "The Lounge",
					body: "🚀 Push notifications have been enabled",
				});
			}
		});

		socket.on("push:unregister", () => client.unregisterPushSubscription(token));
	}

	const sendSessionList = () => {
		const sessions = _.map(client.config.sessions, (session, sessionToken) => ({
			current: sessionToken === token,
			active: _.find(client.attachedClients, (u) => u.token === sessionToken) !== undefined,
			lastUse: session.lastUse,
			ip: session.ip,
			agent: session.agent,
			token: sessionToken, // TODO: Ideally don't expose actual tokens to the client
		}));

		socket.emit("sessions:list", sessions);
	};

	socket.on("sessions:get", sendSessionList);

	if (!Helper.config.public) {
		socket.on("setting:set", (newSetting) => {
			if (!newSetting || typeof newSetting !== "object") {
				return;
			}

			if (typeof newSetting.value === "object" || typeof newSetting.name !== "string" || newSetting.name[0] === "_") {
				return;
			}

			// We do not need to do write operations and emit events if nothing changed.
			if (client.config.clientSettings[newSetting.name] !== newSetting.value) {
				client.config.clientSettings[newSetting.name] = newSetting.value;

				// Pass the setting to all clients.
				client.emit("setting:new", {
					name: newSetting.name,
					value: newSetting.value,
				});

				client.manager.updateUser(client.name, {
					clientSettings: client.config.clientSettings,
				});

				if (newSetting.name === "highlights") {
					client.compileCustomHighlights();
				}
			}
		});

		socket.on("setting:get", () => {
			if (!client.config.hasOwnProperty("clientSettings")) {
				socket.emit("setting:all", {});
				return;
			}

			const clientSettings = client.config.clientSettings;
			socket.emit("setting:all", clientSettings);
		});
	}

	socket.on("sign-out", (tokenToSignOut) => {
		// If no token provided, sign same client out
		if (!tokenToSignOut) {
			tokenToSignOut = token;
		}

		if (!client.config.sessions.hasOwnProperty(tokenToSignOut)) {
			return;
		}

		delete client.config.sessions[tokenToSignOut];

		client.manager.updateUser(client.name, {
			sessions: client.config.sessions,
		});

		_.map(client.attachedClients, (attachedClient, socketId) => {
			if (attachedClient.token !== tokenToSignOut) {
				return;
			}

			const socketToRemove = manager.sockets.of("/").connected[socketId];

			socketToRemove.emit("sign-out");
			socketToRemove.disconnect();
		});

		// Do not send updated session list if user simply logs out
		if (tokenToSignOut !== token) {
			sendSessionList();
		}
	});

	socket.join(client.id);

	const sendInitEvent = (tokenToSend) => {
		socket.emit("init", {
			applicationServerKey: manager.webPush.vapidKeys.publicKey,
			pushSubscription: client.config.sessions[token],
			active: client.lastActiveChannel,
			networks: client.networks.map((network) => network.getFilteredClone(client.lastActiveChannel, lastMessage)),
			token: tokenToSend,
		});
	};

	if (!Helper.config.public && token === null) {
		client.generateToken((newToken) => {
			client.attachedClients[socket.id].token = token = client.calculateTokenHash(newToken);

			client.updateSession(token, getClientIp(socket), socket.request);

			sendInitEvent(newToken);
		});
	} else {
		sendInitEvent(null);
	}
}

function getClientConfiguration(network) {
	const config = _.pick(Helper.config, [
		"public",
		"lockNetwork",
		"displayNetwork",
		"useHexIp",
		"prefetch",
	]);

	config.fileUpload = Helper.config.fileUpload.enable;
	config.ldapEnabled = Helper.config.ldap.enable;

	if (config.displayNetwork) {
		config.defaults = _.clone(network || Helper.config.defaults);
	} else {
		// Only send defaults that are visible on the client
		config.defaults = _.pick(network || Helper.config.defaults, [
			"name",
			"nick",
			"username",
			"password",
			"realname",
			"join",
		]);
	}

	if (!network) {
		config.version = pkg.version;
		config.gitCommit = Helper.getGitCommit();
		config.themes = themes.getAll();
		config.defaultTheme = Helper.config.theme;
		config.defaults.nick = Helper.getDefaultNick();
	}

	if (Uploader) {
		config.fileUploadMaxFileSize = Uploader.getMaxFileSize();
	}

	return config;
}

function getServerConfiguration() {
	const config = _.clone(Helper.config);

	config.stylesheets = packages.getStylesheets();

	return config;
}

function performAuthentication(data) {
	if (typeof data !== "object") {
		return;
	}

	const socket = this;
	let client;
	let token = null;

	const finalInit = () => initializeClient(socket, client, token, data.lastMessage || -1);

	const initClient = () => {
		socket.emit("configuration", getClientConfiguration());

		client.ip = getClientIp(socket);
		client.language = getClientLanguage(socket);

		// If webirc is enabled perform reverse dns lookup
		if (Helper.config.webirc === null) {
			return finalInit();
		}

		reverseDnsLookup(client.ip, (hostname) => {
			client.hostname = hostname;

			finalInit();
		});
	};

	if (Helper.config.public) {
		client = new Client(manager);
		manager.clients.push(client);

		socket.on("disconnect", function() {
			manager.clients = _.without(manager.clients, client);
			client.quit();
		});

		initClient();

		return;
	}

	const authCallback = (success) => {
		// Authorization failed
		if (!success) {
			if (!client) {
				log.warn(`Authentication for non existing user attempted from ${colors.bold(getClientIp(socket))}`);
			} else {
				log.warn(`Authentication failed for user ${colors.bold(data.user)} from ${colors.bold(getClientIp(socket))}`);
			}

			socket.emit("auth", {success: false});
			return;
		}

		// If authorization succeeded but there is no loaded user,
		// load it and find the user again (this happens with LDAP)
		if (!client) {
			client = manager.loadUser(data.user);
		}

		initClient();
	};

	client = manager.findClient(data.user);

	// We have found an existing user and client has provided a token
	if (client && data.token) {
		const providedToken = client.calculateTokenHash(data.token);

		if (client.config.sessions.hasOwnProperty(providedToken)) {
			token = providedToken;

			client.updateSession(providedToken, getClientIp(socket), socket.request);

			return authCallback(true);
		}
	}

	// Perform password checking
	let auth = () => {
		log.error("None of the auth plugins is enabled");
	};

	for (let i = 0; i < authPlugins.length; ++i) {
		if (authPlugins[i].isEnabled()) {
			auth = authPlugins[i].auth;
			break;
		}
	}

	auth(manager, client, data.user, data.password, authCallback);
}

function reverseDnsLookup(ip, callback) {
	dns.reverse(ip, (err, hostnames) => {
		if (!err && hostnames.length) {
			return callback(hostnames[0]);
		}

		callback(ip);
	});
}
